File: Searching/374_Guess_Number_Higher_or_Lower/solution374.cpp
int guess(int num);

class Solution {
public:
    int guessNumber(int n) {
        int low = 1, high = n, mid;
        
        while (true) {
            mid = low + (high - low) / 2;
            int res = guess(mid);
            
            if (res == 0) {
                return mid;
            } else if (res == -1) {
                high = mid - 1;
            } else {
                low = mid + 1;
            }
        }
    }
};

// Runtime1: 0ms, beats 100%
// Memory1: 7.40MB, beats 98.44%

// Runtime2: 0ms, beats 100%
// Memory2: 7.67MB, beats 15.49%

// Runtime3: 0ms, beats 100%
// Memory3: 7.59MB, beats 69.62%

File: Searching/374_Guess_Number_Higher_or_Lower/solution374.py
class Solution:
    def guessNumber(self, n: int) -> int:
        low, high = 1, n
        while (res := guess(mid := ((high + low) >> 1))) != 0:
            if res == -1: high = mid - 1
            else: low = mid + 1
        return mid

# Runtime1: 37ms, beats 38.08%
# Memory1: 16.45MB, beats 68.11%

# Runtime2: 27ms, beats 92.82%
# Memory2: 16.36MB, beats 94.17%

# Runtime3: 23ms, beats 98.27%
# Memory3: 16.37MB, beats 94.17%